File: src/backup.php
<?php
// Configuration
$api_url = getenv('API_URL') . "/invoices";
$auth_header = getenv('API_KEY');
$backup_dir = './backups';
$file_age_limit = 14 * 86400; // 14 days in seconds

echo "API URL: " . $api_url . "\n";
echo "API KEY: " . $auth_header . "\n";

// Ensure backup directory exists
if (!is_dir($backup_dir) && !mkdir($backup_dir, 0755, true) && !is_dir($backup_dir)) {
  die('Failed to create backup directory.');
}

// Prepare the backup file path
$file_name = sprintf('%s/%s.json', $backup_dir, date('d-m-Y-H:i'));

// Fetch API data
$ch = curl_init($api_url);
curl_setopt_array($ch, [
  CURLOPT_RETURNTRANSFER => true,
  CURLOPT_HTTPHEADER => ['Authorization: ' . $auth_header]
]);
$response = curl_exec($ch);
$http_code = curl_getinfo($ch, CURLINFO_HTTP_CODE);
curl_close($ch);

echo $response;

// Handle HTTP response
if ($http_code !== 200) {
  $error_messages = [
    401 => 'Unauthorized. Check your authorization header.',
    404 => 'API endpoint not found.',
    500 => 'Internal Server Error. There may be an issue with the API.'
  ];
  echo $error_messages[$http_code] ?? "Error: HTTP status code $http_code" . PHP_EOL;
  exit;
}

// Save response to JSON file
file_put_contents($file_name, $response);
echo "Data saved to $file_name" . PHP_EOL;

// Clean up old backup files
foreach (new DirectoryIterator($backup_dir) as $file) {
  if ($file->isFile() && (time() - $file->getMTime() > $file_age_limit)) {
    unlink($file->getPathname());
    echo 'Deleted old file: ' . $file->getFilename() . PHP_EOL;
  }
}

File: src/backup.py
import os
import requests
import json
from datetime import datetime, timedelta

# Configuration
# API_URL = os.getenv('API_URL') + "/api/invoices"
# API_KEY = os.getenv('API_KEY')
API_URL = "https://invoice-server-main-yvc4uq.laravel.cloud/api/invoices"
API_KEY = "GKLdm2392J38iFgGPjcOOz2iQ327"
BACKUP_DIR = './backups'
FILE_AGE_LIMIT = 14  # days


def fetch_data():
    headers = {'Authorization': API_KEY}
    response = requests.get(API_URL, headers=headers)
    if response.status_code == 200:
        return response.json()
    else:
        raise Exception(f"Error: HTTP status code {response.status_code}")


def save_backup(data):
    os.makedirs(BACKUP_DIR, exist_ok=True)
    file_name = f"{BACKUP_DIR}/{datetime.now().strftime('%d-%m-%Y-%H-%M')}.json"
    with open(file_name, 'w') as file:
        json.dump(data, file)
    print(f"Data saved to {file_name}")


def clean_old_backups():
    limit = datetime.now() - timedelta(days=FILE_AGE_LIMIT)
    for file in os.listdir(BACKUP_DIR):
        file_path = os.path.join(BACKUP_DIR, file)
        if os.path.isfile(file_path) and datetime.fromtimestamp(os.path.getmtime(file_path)) < limit:
            os.remove(file_path)
            print(f"Deleted old file: {file}")


if __name__ == "__main__":
    try:
        data = fetch_data()
        print(data)
        save_backup(data)
        clean_old_backups()
    except Exception as e:
        print(e)
